authorization/service/deleteDevices.go

```diff
@@ -0,0 +1,135 @@
+package service
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	grpc_auth "github.com/grpc-ecosystem/go-grpc-middleware/auth"
+	"github.com/plgd-dev/cloud/authorization/events"
+	"github.com/plgd-dev/cloud/authorization/pb"
+	"github.com/plgd-dev/cloud/authorization/persistence"
+	"github.com/plgd-dev/cloud/pkg/log"
+	"github.com/plgd-dev/cloud/pkg/net/grpc"
+	pkgTime "github.com/plgd-dev/cloud/pkg/time"
+	"github.com/plgd-dev/cloud/resource-aggregate/cqrs/utils"
+	"github.com/plgd-dev/kit/strings"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func getUniqueDeviceIds(deviceIds []string) []string {
+	devices := make(strings.Set)
+	devices.Add(deviceIds...)
+	delete(devices, "")
+	return devices.ToSlice()
+}
+
+func getUserDevices(tx persistence.PersistenceTx, owner string) ([]string, error) {
+	it := tx.RetrieveByOwner(owner)
+	defer it.Close()
+	if it.Err() != nil {
+		return nil, fmt.Errorf("failed to obtain user devices: %w", it.Err())
+	}
+	var deviceIds []string
+	var d persistence.AuthorizedDevice
+	for it.Next(&d) {
+		deviceIds = append(deviceIds, d.DeviceID)
+	}
+	return deviceIds, nil
+}
+
+func (s *Service) publishDevicesUnregistered(ctx context.Context, owner string, deviceIDs []string) error {
+	v := events.Event{
+		Type: &events.Event_DevicesUnregistered{
+			DevicesUnregistered: &events.DevicesUnregistered{
+				Owner:     owner,
+				DeviceIds: deviceIDs,
+				AuditContext: &events.AuditContext{
+					UserId: owner,
```

Comment:
This should be always the "sub" claim. Config.ownerClaim can be changed from "sub" to something else and in such a case, we lose the information on who did this change.

Reply:
Shouldn't it be removed from configuration then if it shouldn't be changed?

Reply:
No, because the `event.Owner` should match `ownerClaim` claim value. `AuditContext` always "sub". Reason:
`ownerClaim` is set to tenantId - organization identifier. That means that the device is owned by the organization and any employee of the organization can retrieve it. If you pass the same id (tenantId) to the AuditContext, all you can track back is that this organization deleted it. But you don't know which bad guy did this mistake, because you didn't take his userId, which is always in "sub" claim :) So it's a good feature for bad guys, but not for bad security managers :D

Reply:
In the case that ownerClaim != 'sub' where will both claims be. In the JTW? Is it guaranteed to always have both non-empty? Or what should I do when 'sub' or custom claim is empty?

Reply:
All gateways have to verify 
if ownerClaim == "sub"
  jwt.sub must be present
if ownerClaim == "abc"
  jwt.sub && jwt.abc must be present

Reply:
Will be covered in the separate task. @Danielius1922 please add todo at this line

Reply:
TODO added